sklearn/cluster/tests/test_affinity_propagation.py

```diff
@@ -181,6 +181,26 @@ def test_equal_similarities_and_preferences():
     assert _equal_similarities_and_preferences(S, np.array(0))
 
 
+def test_affinity_propagation_random_status():
+    # Significance of random_state parameter
+    rng = np.random.RandomState(42)
+    X = rng.rand(40, 10)
+    y = (4 * rng.rand(40)).astype(np.int)
+
+    # random_state = 0
+    ap = AffinityPropagation(convergence_iter=1, max_iter=2, random_state=0)
+    ap.fit(X, y)
+    centers0 = ap.cluster_centers_
+
+    # random_state = 76
+    ap = AffinityPropagation(convergence_iter=1, max_iter=2, random_state=76)
+    ap.fit(X, y)
+    centers76 = ap.cluster_centers_
+
+    centers_diff = centers0 - centers76
+    assert not np.array_equal(centers_diff, np.zeros(len(centers_diff)))
```

Comment:
I think you could just use `0` instead of an array.
However it might be better to check a lower bound on the (potentially squared) difference

Reply:
> I think you could just use `0` instead of an array.

you are right .. I really wasn't aware of that! Thanks for the trick... there is some implicit conversion there I suppose... yet, may I still use the array version? I feel as it is less confusing to explicit state that we are comparing two arrays...
   
> However it might be better to check a lower bound on the (potentially squared) difference

May I ask why? Thanks for your patience.

Reply:
> there is some implicit conversion there I suppose

Indeed, it's array [broadcasting](https://docs.scipy.org/doc/numpy/user/theory.broadcasting.html#array-broadcasting-in-numpy). It's the same as when you do e.g. `out = some_array + 3` or `normalized = some_array / some_array.sum()`

> may I still use the array version

sure

> > However it might be better to check a lower bound on the (potentially squared) difference

> May I ask why?

because right now you're just checking that the diff is non-zero. But it could be extremely close to zero, and we wouldn't know by how much. Asserting a lower bound makes sure that the difference is `at least x`, where `x` is hopefully significantly non-zero.

Reply:
Thanks for your answer. I'm probably missing what I need to test here... my understanding was that I should check that two different initial states produce different results in the early stages of the computation. Why do I need to quantify this difference? How do I define an objective significant lower limit?

Reply:
> my understanding was that I should check that two different initial states produce different results in the early stages of the computation

Indeed, but in general, comparing float with exact values (as done by `np.array_equal`) isn't recommended. That's because of floating point issues which may happen for whatever reason. Consider this:

```py
In [2]: a = np.arange(5)                                                                                               
In [4]: np.array_equal(a, a + 1e-20)                                                                                   
Out[4]: False
```

Are the 2 arrays really that different?

> Why do I need to quantify this difference?

It's a guarantee for us that the difference is significant and that the test doesn't pass just by luck

> How do I define an objective significant lower limit?

Pretty much arbitrary (in this case it's gonna depend on the scale of X). Floating errors are sometimes in the range of 1e-6 depending on the precision. I would say if you have a difference of more than `1e-2` that's fine, though the higher the better.

BTW, it looks like both centers vectors are actually empty lists

Reply:
> BTW, it looks like both centers vectors are actually empty lists

Aaahhhh! Let me check! :(